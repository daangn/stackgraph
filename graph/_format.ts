// 1.40.5: https://github.com/denoland/deno/issues/22496
import { stripAnsiCode } from "https://deno.land/std@0.216.0/fmt/colors.ts"
import { Reducer, Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { Declaration, DeclDeps } from "./decl_deps.ts"
import { encodeVSCodeURI, prettyPrintURI } from "./vscode_uri.ts"

export const serializeNoColor = (x: unknown) =>
	stripAnsiCode(
		Deno.inspect(x, {
			depth: Infinity,
			colors: false,
			sorted: true,
			trailingComma: true,
			compact: false,
			iterableLimit: Infinity,
			breakLength: Infinity,
			escapeSequences: false,
			strAbbreviateSize: Infinity,
		}),
	)

export const asRecord =
	<T extends string | number | symbol>(fn: (decl: Declaration) => T) =>
	(declDeps: DeclDeps): Record<T, Set<T>> =>
		Stream.from(declDeps.entries())
			.map(([decl, deps]) => [fn(decl), new Set(deps.map(fn))] as [T, Set<T>])
			.reduce(Reducer.toJSObject())

export const declDepsSerializer = (declDeps: DeclDeps) =>
	serializeNoColor(
		asRecord((x) => prettyPrintURI(encodeVSCodeURI(x)))(declDeps),
	)
