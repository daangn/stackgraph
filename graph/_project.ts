import { Reducer, Stream } from "../deps/rimbu.ts"
import { Project, SourceFile } from "../deps/ts_morph.ts"
import { denoCompilerOptions } from "../utils/project.ts"

/** In-Memory ESM Project for quick testing */
export const inMemoryProject = () =>
	new Project({
		compilerOptions: denoCompilerOptions,
		useInMemoryFileSystem: true,
	})

/**
 * Add source files to in-memory project
 */
export const withSrc =
	(project: Project) =>
	<const T extends string>(src: Record<T, string>): Record<T, SourceFile> =>
		Stream.fromObject(src)
			.map(([path, text]) =>
				[path, project.createSourceFile(path, text)] as const
			)
			.reduce(Reducer.toJSObject())
