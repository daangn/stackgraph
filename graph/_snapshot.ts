const read = await Deno.permissions.query({ name: "read" })

const ignore = read.state !== "granted"

/**
 * same as `Deno.test` but skipped if `read` permission is denied
 */
export const snapshotTest = (
	name: Deno.TestDefinition["name"],
	fn: Deno.TestDefinition["fn"],
) => Deno.test({ name, ignore, fn })
