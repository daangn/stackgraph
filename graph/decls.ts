import {
	Stream,
	StreamSource,
} from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { SourceFile } from "../deps/ts_morph.ts"
import { Declaration } from "./decl_deps.ts"

/**
 * Find initial declarations to begin analyze with.
 */
export const getDecls =
	(filter: (node: Declaration) => boolean) =>
	(files: StreamSource<SourceFile>): Stream<Declaration> =>
		Stream.from(files).flatMap((file) =>
			Stream.from<Declaration>(
				file.getVariableDeclarations(),
				file.getFunctions(),
				file.getClasses(),
			).filter(filter)
		)

export const asMetadataRecord = <A, B>(links: LinkedDecl<A, B>[]) =>
	Object.fromEntries(
		links.map(({ node, ...metadata }) => [node.getName()!, metadata]),
	)
