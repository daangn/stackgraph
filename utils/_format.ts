import { type DeclDeps } from "../graph/decl_deps.ts"

/**
 * prints TS AST node names
 */
export const printNodes = (...data: unknown[]) => {
	const xs = data.map((x) => {
		if (!x || typeof x !== "object") return x
		if (Array.isArray(x)) return x.map((x) => x?.getText?.() ?? x)
		// deno-lint-ignore no-explicit-any
		return (x as any).getText?.() ?? x
	})
	console.log(...xs)
}

export const prettyPrintDeclDeps = (declDeps: DeclDeps) =>
	declDeps.forEach((deps, decl) => {
		printNodes(decl, "->", deps)
	})
