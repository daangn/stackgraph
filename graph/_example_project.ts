import { Stream } from "https://deno.land/x/rimbu@1.2.0/stream/mod.ts"
import { inMemoryProject } from "./_project.ts"

/**
 * https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements
 *
 * Example file of all top-level javascript declarations
 */
export const declExampleText = (h: string | number = "") => /*javascript*/ `
    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#declaring_variables
    export var varVar${h} = 1
    export var varVarIgnore${h} = 2

    export let varLet${h} = 1
    export let varLetIgnore${h} = 2

    export const varConst${h} = 1
    export const varConstIgnore${h} = 2

    export const varArrowFn${h} = () => {}
    export const varArrowFnIgnore${h} = () => {}

    // https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements#functions_and_classes
    export function functionFn${h}() {}
    export function functionFnIgnore${h}() {}

    export function${h}* functionGen() {}
    export function${h}* functionGenIgnore() {}

    export async function${h} functionAsync() {}
    export async function${h} functionAsyncIgnore() {}

    export async function${h}* functionAsyncGen() {}
    export async function${h}* functionAsyncGenIgnore() {}

    export class Class${h} {}
    export class ClassIgnore${h} {}

    // others
    if (true) {
        const variableInIf${h} = 3
    }
`

/**
 * Arbitarily large file with many declarations for benchmarking
 */
export const getDeclExampleFile = (amount: number) => {
	const project = inMemoryProject()

	const text = Stream
		.range({ amount }).map(declExampleText).join({ sep: "\n\n" })

	return project.createSourceFile("a.ts", text)
}

export const exampleSrc = {
	"a.ts": /*javascript*/ `
        export const a = "a"

	    export const aaa = () => a
	`,
	"b.tsx": /*javascript*/ `
        import { a, aaa } from "./a.ts"
        export { a }

        export const Comp = () => <div>{a}</div>
        export const CompAAA = () => <div>{aaa()}</div>

        // unrelated to a and should be ignored
        export const Unrelated = () => <div>Unrelated</div>
    `,
	"c.tsx": /*javascript*/ `
        import { Comp } from "./b.tsx"

        export const Page = () => <div><Comp/></div>
    `,
	"d.tsx": /*javascript*/ `
        import { a } from "./b.tsx"
        import { Page } from "./c.tsx"

        export const InnerImport = () => <Page/>
        export const AliasedImport = () => <div>{a}</div>
    `,
}
