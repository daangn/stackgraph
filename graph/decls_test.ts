import type { Declaration } from "./decl_deps.ts"
import { getDecls } from "./decls.ts"
import { declExampleText } from "./_example_project.ts"
import { inMemoryProject } from "./_project.ts"
import { assertSnapshot } from "../test_deps.ts"
import { serializeNoColor } from "./_format.ts"
import { snapshotTest } from "./_snapshot.ts"

snapshotTest("getInitialDecls() filters by declaration name", async (t) => {
	const project = inMemoryProject()

	const file = project.createSourceFile("a.ts", declExampleText())
	const filter = (node: Declaration) =>
		!node.getName()?.toLowerCase().endsWith("ignore")

	const decls = getDecls(filter)([file]).map((x) => x.getName()).toArray()

	await assertSnapshot(t, decls, { serializer: serializeNoColor })
})
