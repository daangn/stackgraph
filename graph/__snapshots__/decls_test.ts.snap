export const snapshot = {};

snapshot[`getInitialDecls() filters by declaration name 1`] = `
[
  "varVar",
  "varLet",
  "varConst",
  "varArrowFn",
  "functionFn",
  "functionGen",
  "functionAsync",
  "functionAsyncGen",
  "Class",
]
`;
